pkg/controller/serviceexternalip/controller.go

```diff
@@ -240,32 +303,95 @@ func (c *ServiceExternalIPController) processNextWorkItem() bool {
 	return true
 }
 
-func (c *ServiceExternalIPController) releaseExternalIP(service apimachinerytypes.NamespacedName) error {
+func (c *ServiceExternalIPController) releaseExternalIP(service apimachinerytypes.NamespacedName) (bool, error) {
 	c.ipAllocationMutex.Lock()
 	defer c.ipAllocationMutex.Unlock()
-	allocation, exists := c.ipAllocationMap[service]
+	obj, exists, _ := c.ipAllocations.GetByKey(service.String())
 	if exists {
-		if err := c.externalIPAllocator.ReleaseIP(allocation.ipPool, allocation.ip); err != nil {
-			if err == externalippool.ErrExternalIPPoolNotFound {
-				// Ignore the error since the external IP Pool could be deleted.
-				klog.Warningf("Failed to release IP %s for service %s: IP Pool %s does not exist", service, allocation.ip, allocation.ipPool)
+		allocation := obj.(*ipAllocation)
+		objs, _ := c.ipAllocations.ByIndex(ipIndex, allocation.ip.String())
+		// The IP is exclusively used by the Service.
+		if len(objs) == 1 {
+			if err := c.externalIPAllocator.ReleaseIP(allocation.ipPool, allocation.ip); err != nil {
+				if err == externalippool.ErrExternalIPPoolNotFound {
+					// Ignore the error since the external IP Pool could be deleted.
+					klog.Warningf("Failed to release IP %s for service %s: IP Pool %s does not exist", service, allocation.ip, allocation.ipPool)
+				} else {
+					klog.ErrorS(err, "Failed to release external IP", "service", service, "ip", allocation.ip, "pool", allocation.ipPool)
+					return false, err
+				}
 			} else {
-				klog.ErrorS(err, "Failed to release external IP", "service", service, "ip", allocation.ip, "pool", allocation.ipPool)
-				return err
+				klog.InfoS("Released external IP", "service", service, "ip", allocation.ip, "pool", allocation.ipPool)
 			}
-		} else {
-			klog.InfoS("Released external IP", "service", service, "ip", allocation.ip, "pool", allocation.ipPool)
 		}
-		delete(c.ipAllocationMap, service)
+		c.ipAllocations.Delete(obj)
+		// Update other Services as they may be affected by the release of this IP.
+		c.enqueueServicesWithoutIPs(allocation.ipPool)
```

Comment:
This is for this case: a service sharing an IP with other Services changes its annotation from shareable to unsharable, in which case we didn't release its IP as I want to limit the annotation for requesting IP only, at the moment new Services can no longer get this IP. After deleting this Service, new Services can get this IP. Maybe I should add a comment for it.

Reply:
Got it, the new comment helps. This is because we can get in situations where Services are sharing the same IP even though not all Services have the annotation, based on user actions.